Add slice support for FileRegion.

Apache9 opened the pull request:
Motivation:
Let codec such as HTTP/2 can write FileRegion chunk by chunk.

Modifications:
Add a transferIndex which is similar to readerIndex in ByteBuf for FileRegion.
Add two transferBytesTo methods which are similar to readBytes in ByteBuf for FileRegion.
Add slice and transferSlice methods which are similar to slice and readSlice in ByteBuf for FileRegion.
Make the return value of position to be the offset related to the FileChannel returned by channel method. And for DefaultFileRegion, the return value of the original position method is the offset related to the underlying FileChannel, so this does not break compatibility.
The new methods are added to a SliceableFileRegion interface which extends the old FileRegion interface, will move the new methods to FileRegion at the next minor release.
Add a FileChannelBasedFileRegion which makes it easier for implementation file channel related file regions.
Add a FileChannelBasedSlicedFileRegion which is package private.
Add FileChannelBasedSlicedFileRegionTest.
The native transport has not been modified yet so it can not accept the FileChannelBasedSlicedFileRegion right now, will address it in another PR.

Result:
Passed all FileRegion related testsuites.

Fixes
The first step of #3927

Apache9 commented:
The first step of #3927

Apache9 commented:
ping @Scottmitch @nmittler @normanmaurer . Thanks.

normanmaurer commented on transport/src/main/java/io/netty/channel/FileRegion.java:
```diff
@@ -84,9 +86,86 @@
+     *                  byte of the region transferred.
+     * @param length    the maximum number of bytes to transfer
+     */
+    long transferBytesTo(WritableByteChannel target, long position, long length) throws IOException;
```
We can not add methods to the interface until we release a new minor / major version. So please remove this and just check for the actual implementation with an instanceof when you need these methods.

normanmaurer replied:
This is true for all new methods on the interface

Apache9 replied:
OK. So it is free for me to add a new AbstractFileRegion class and make DefaultFileRegion implement it, and until netty 4.2 release I can move the new methods in AbstractFileRegion up to the interface?

normanmaurer replied:
yep I think this is fine. (Please also add a todo to move to interface at some point.

Apache9 replied:
Got it. Thanks for the explaination. Will fix this soon.

normanmaurer commented on transport/src/main/java/io/netty/channel/FileRegion.java:
```diff
@@ -84,9 +86,86 @@
+    /**
+     * Return the underlying file channel instance of this file region.
+     */
+    FileChannel channel() throws IOException;
```
see above also I think this should be not part of the `FileRegion` interface at all as the FileRegion should not be specific to `FileChannel`.

normanmaurer commented on transport/src/main/java/io/netty/channel/DefaultFileRegion.java:
```diff
@@ -168,24 +183,55 @@ protected void deallocate() {
     @Override
-    public FileRegion retain(int increment) {
-        super.retain(increment);
+    public DefaultFileRegion retain(int increment) {
```
we can not change the return type as this would be a breaking change.

Apache9 replied:
One more question, is this because we may break the code which extends DefaultFileRegion, or that we want to maintain drop in replacement between patch releases of netty since the JVM may include the return value to generate the method signature? I ask this because the Native.sendfile uses DefaultFileRegion as the parameter and I want to change it to AbstractFileRegion if possible in the follow on PRs which add native support for SlicedFileRegion.

normanmaurer replied:
its because the user may extend it.

Apache9 replied:
Got it. Thanks.

normanmaurer commented on transport/src/main/java/io/netty/channel/DefaultFileRegion.java:
```diff
@@ -168,24 +183,55 @@ protected void deallocate() {
 
     @Override
-    public FileRegion touch(Object hint) {
+    public DefaultFileRegion touch(Object hint) {
```
we can not change the return type as this is a breaking change

normanmaurer commented on transport/src/test/java/io/netty/channel/SlicedFileRegionTest.java:
```diff
@@ -0,0 +1,105 @@
+ */
+public class SlicedFileRegionTest {
+
+    private static byte[] TEST_DATA = new byte[1024 * 64];
```
final

normanmaurer commented on transport/src/test/java/io/netty/channel/SlicedFileRegionTest.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Test
+    public void test() throws IOException {
```
use a more meaningful name.

normanmaurer commented on transport/src/main/java/io/netty/channel/SlicedFileRegion.java:
```diff
@@ -0,0 +1,122 @@
+/**
+ * A derived {@link FileRegion} which shares reference counting with its parent.
+ */
+public class SlicedFileRegion extends AbstractFileRegion {
```
make this final and package private for now as its an implementation detail.

normanmaurer commented on transport/src/main/java/io/netty/channel/FileRegion.java:
```diff
@@ -84,9 +86,86 @@
+     * <p>
+     * Also be aware that this method will NOT call {@link #retain()} and so the reference count will NOT be increased.
+     */
+    FileRegion slice(long index, long length);
```
can you explain what the slices are good for ?

Apache9 replied:
For splitting big chunk into data frames in HTTP/2 implementation.

normanmaurer replied:
ok.

Apache9 commented:
@normanmaurer I've modified the code based on your comments. Now the new methods are only in AbstractFileRegion. In the XXXByteStreamChannel implementations I've added a new branch to handle AbstractFileRegion and also added a comment to say that please remove the branch once we release netty 4.2 and move the methods in AbstractFileRegion to FileRegion.

netkins commented on transport/src/main/java/io/netty/channel/oio/OioByteStreamChannel.java:
```diff
@@ -128,21 +129,44 @@ protected void doWriteFileRegion(FileRegion region) throws Exception {
-            if (written >= region.count()) {
-                return;
+        } else {
+            // TODO: consider remove this branch once we move the methods in AbstractFileRegion
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

netkins commented on transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java:
```diff
@@ -206,7 +207,42 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+                    break;
+                }
             } else if (msg instanceof FileRegion) {
+                // TODO: consider remove this branch once we move the methods in AbstractFileRegion
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

netkins commented on transport/src/main/java/io/netty/channel/socket/nio/NioSocketChannel.java:
```diff
@@ -380,8 +381,15 @@ protected int doWriteBytes(ByteBuf buf) throws Exception {
+            AbstractFileRegion r = (AbstractFileRegion) region;
+            return r.transferBytesTo(javaChannel(), r.transferableBytes());
+        } else {
+            // TODO: consider remove this branch once we move the methods in AbstractFileRegion
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

netkins commented on transport/src/main/java/io/netty/channel/oio/AbstractOioByteChannel.java:
```diff
@@ -203,7 +204,15 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+                in.progress(region.transferIndex() - transferIndex);
+                in.remove();
             } else if (msg instanceof FileRegion) {
+                // TODO: consider remove this branch once we move the methods in AbstractFileRegion
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

Apache9 commented:
Any other concerns?  @normanmaurer Thanks.

normanmaurer commented on transport/src/main/java/io/netty/channel/AbstractFileRegion.java:
```diff
@@ -0,0 +1,184 @@
+     * Returns {@code true} if and only if {@link #transferableBytes()} is greater than {@code 0}.
+     */
+    public boolean isTransferable() {
+        return transferIndex < count();
```
use `transferableBytes() > 0` so the impl match what you describe in the javadocs

Apache9 replied:
I just follow the implementation of AbstractByteBuf.  `readableBytes` returns `writerIndex - readerIndex` while `isReadable` returns `writerIndex > readerIndex`. I'm OK with using `transferableBytes() > 0`, just want to make sure that there is no other reasons why ByteBuf use `writerIndex > readerIndex` instead of `readableBytes() > 0`

Apache9 replied:
I changed the javadoc to 'Returns {@code true} if and only if {@code (this.count - this.transferIndex)} is greater than {@code 0}.'

normanmaurer commented on transport/src/main/java/io/netty/channel/AbstractFileRegion.java:
```diff
@@ -0,0 +1,184 @@
+        }
+        length = Math.min(count() - position, length);
+        if (length == 0) {
+            return 0L;
```
nit: You can remove the `L`

normanmaurer commented on transport/src/main/java/io/netty/channel/AbstractFileRegion.java:
```diff
@@ -0,0 +1,184 @@
+     * Also be aware that this method will NOT call {@link #retain()} and so the reference count
+     * will NOT be increased.
+     */
+    public AbstractFileRegion slice(long index, long length) {
```
This should return a FileRegion imho.

Apache9 replied:
Here we need the returned FileRegion also has the newly added method, otherwise we need to do instanceof and cast everywhere after a slice. We can change to `FileRegion` after moving the methods to FileRegion interface?

normanmaurer commented on transport/src/main/java/io/netty/channel/AbstractFileRegion.java:
```diff
@@ -0,0 +1,184 @@
+     * Note that this method returns a {@linkplain #retain() retained} file region unlike
+     * {@link #slice()}.
+     */
+    public AbstractFileRegion retainedSlice(long index, long length) {
```
This should return a `FileRegion` imho

normanmaurer commented on transport/src/main/java/io/netty/channel/AbstractFileRegion.java:
```diff
@@ -0,0 +1,184 @@
+     * @param length
+     *            the size of the new slice
+     */
+    public AbstractFileRegion transferSlice(long length) {
```
This should return a `FileRegion` imho

normanmaurer commented on transport/src/main/java/io/netty/channel/AbstractFileRegion.java:
```diff
@@ -0,0 +1,184 @@
+     * @param length
+     *            the size of the new slice
+     */
+    public AbstractFileRegion transferRetainedSlice(long length) {
```
This should return a `FileRegion` imho

normanmaurer commented on transport/src/main/java/io/netty/channel/AbstractFileRegion.java:
```diff
@@ -0,0 +1,184 @@
+     *
+     * @return {@code null} if this file region is not a wrapper
+     */
+    public abstract AbstractFileRegion unwrap();
```
This should return a `FileRegion` imho

normanmaurer commented on transport/src/main/java/io/netty/channel/AbstractFileRegion.java:
```diff
@@ -0,0 +1,184 @@
+    /**
+     * Return the underlying file channel instance of this file region.
+     */
+    public abstract FileChannel channel() throws IOException;
```
I really dislike that we need to expose this at all as a FileRegion should not really depend on a FileChannel. This is an implementation detail.

Apache9 replied:
Yeah I think we do not need to move this method to the FileRegion interface. This is just used to simplify the implementation of sub classes, and also for native transport such as epoll and kqueue. So maybe rename AbstractFileRegion to FileChannelBasedFileRegion?

normanmaurer commented on transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java:
```diff
@@ -184,7 +185,7 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
                     writeSpinCount = config().getWriteSpinCount();
                 }
-                for (int i = writeSpinCount - 1; i >= 0; i --) {
+                for (int i = writeSpinCount - 1; i >= 0; i--) {
```
revert

Apache9 commented:
Introduce a new interface called SliceableFileRegion. The new methods are all added to this interface. And also introduce a FileChannelBasedFileRegion class to place the common code. The `FileChannel channel()` method is marked as protected since in JNI seems we are free to call protected methods.

@normanmaurer  PTAL. Thanks.

Apache9 commented:
Fix checkstyle problem.

Ping @normanmaurer .

Thanks.

netkins commented on transport/src/main/java/io/netty/channel/oio/OioByteStreamChannel.java:
```diff
@@ -128,21 +129,44 @@ protected void doWriteFileRegion(FileRegion region) throws Exception {
-            if (written >= region.count()) {
-                return;
+        } else {
+            // TODO: consider remove this branch once we move the methods in SliceableFileRegion
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

netkins commented on transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java:
```diff
@@ -206,7 +207,42 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+                    break;
+                }
             } else if (msg instanceof FileRegion) {
+                // TODO: consider remove this branch once we move the methods in SliceableFileRegion
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

Scottmitch replied:
can you create a followup issue for this, and link it to this PR?

netkins commented on transport/src/main/java/io/netty/channel/socket/nio/NioSocketChannel.java:
```diff
@@ -380,8 +381,15 @@ protected int doWriteBytes(ByteBuf buf) throws Exception {
+            SliceableFileRegion r = (SliceableFileRegion) region;
+            return r.transferBytesTo(javaChannel(), r.transferableBytes());
+        } else {
+            // TODO: consider remove this branch once we move the methods in SliceableFileRegion
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

netkins commented on transport/src/main/java/io/netty/channel/oio/AbstractOioByteChannel.java:
```diff
@@ -203,7 +204,15 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+                in.progress(region.transferIndex() - transferIndex);
+                in.remove();
             } else if (msg instanceof FileRegion) {
+                // TODO: consider remove this branch once we move the methods in SliceableFileRegion
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

normanmaurer commented:
@ejona86 @carl-mastrangelo @Scottmitch  PTAL as well

ejona86 commented on transport/src/main/java/io/netty/channel/SliceableFileRegion.java:
```diff
@@ -0,0 +1,131 @@
+/**
+ * A {@link FileRegion} which supports slice.
+ * <p>
+ * We introduce this interface is only because that we can not add new methods to an interface until
```
"is only because that" seems to have a grammar issue. Maybe: "We introduce this interface because we..."

Apache9 replied:
Done.

ejona86 commented on transport/src/main/java/io/netty/channel/FileChannelBasedSlicedFileRegion.java:
```diff
@@ -0,0 +1,133 @@
+
+    @Override
+    public long transferTo(WritableByteChannel target, long position) throws IOException {
+        throw new UnsupportedOperationException();
```
If `transferTo()` and `transferred()` are unsupported, then what's the point of `SliceableFileRegion` extending `FileRegion`? It's simply impossible to use this `SliceableFileRegion` as a `FileRegion`. They could be separate interfaces and although `DefaultFileRegion` could implement both, the slices from it would only implement `SliceableFileRegion`.

ejona86 replied:
Why isn't `transferTo()` supported? What would be wrong with this sketch of an implementation:
```java
private final long transferred;

public long transferred() { return transferred; }
￼
public long transferTo(WritableByteChannel target, long position) throws IOException {
  long written = transferBytesTo(target, position, this.count - position);
  this.transferred += written;
  return written;
}
```

Apache9 replied:
For now `SliceableFileRegion` just extends `FileRegion`, the intention here is to move the methods in this interface up to `FileRegion` at the next minor release. The old `transferTo` and `trnsferred` methods are marked as deprecated and some new methods are provided. You can either use the old methods, or the new methods, but you can not use them both at the same time. So you mean it is better to declare a completely new interface and remove the old `FileRegion` interface in the future?

Apache9 replied:
These methods are expected to be removed in the future and new implementation classes are not required to implement them. If you use the new methods introduced in `SliceableFileRegion` then you need to use the new `transferBytesTo` methods. Of course this is only my expectation, if this is not the correct upgrading method of netty, please tell me. I can modify the code to respect the netty rule. Thanks.

ejona86 replied:
> You can either use the old methods, or the new methods

Who do you mean by "you". _I_ can't use a method that throws `UnsupportedOperationException`. So if I am given an instance of `FileRegion`, I don't know if I can use it. The important detail is the the _creator_ of the instance is different than the _consumer_ of the instance.

> but you can not use them both at the same time

That's not what's implemented. In the current implementation the _consumer_ doesn't get a choice. As far the API goes, the consumer is the only one that matters; the creator knows the concrete type and so has little need for interfaces.

> So you mean it is better to declare a completely new interface and remove the old FileRegion interface in the future?

With the current implementation, yes. But you could also change the implementation.

Apache9 replied:
The only way to create a `SlicedFileRegion` is to use the methods declared in `SliceableFileRegion`, then you also need to use the `transferBytesTo` methods declared in `SliceableFileRegion`.  And you can still use `DefaultFileRegion` in the old way. That's what I mean. Sorry I do not speak English very well.

ejona86 replied:
> These methods are expected to be removed in the future

So this argument feels like "because we're lazy," because the methods _are_ there today. "Let's not spend the effort of implementing them if they will be removed later." Which is a fair argument, but not by itself.

My implementation sketch would seem to indicate it isn't very hard to add support for the old method. Some edge cases might need to be checked, but that still seems to be pretty easy. So assuming my proposed implementation works, the "lazy" here seems to be quite extreme, as we don't want to add ~10 straight-forward lines to maintain compatibility. But I'd accept that maybe there's an issue with my implementation, which could dramatically change the cost/benefit.

> new implementation classes are not required to implement them.

This is a circular argument. I'm claiming that implementations should implement them, and that I don't see a strong reason why they shouldn't implement them.

> If you use the new methods introduced in SliceableFileRegion then you need to use the new transferBytesTo methods.

As I mention in the other comment, the problem is that these are two different "you"s. The code creating the instance is different than the one consuming it, otherwise they wouldn't need to communicate via an interface.

ejona86 replied:
> The only way to create a SlicedFileRegion is to use the methods declared in SliceableFileRegion

Not quite, since `DefaultFileRegion.slice(long, long)` is concrete, but that's a technicality. You're right that anyone slicing probably doesn't have the concrete type; I was wrong there. But I don't think it really changes my point that the producer and consumer are different pieces of code.

> Sorry I do not speak English very well.

Oh, wow. I didn't realize you were a non-native speaker. Your English is great!

Apache9 replied:
I just want to follow the `ByteBuf` way to make the code a bit easier to understand when adding the slice support. And to be honest I'm also a user of netty and I just want this feature, no matter who implement it. So if you think the code is BS and you already have an implementation then just implement it in netty. Or you can just modify my code and implement the compatibility code. I do not mind. Thanks.

ejona86 replied:
I don't think your code should get thrown out. I think there are two easy alternatives to address the problem I brought up:

1. Implement `transferTo()`. It seems easy to implement, so why not? This lets pre-existing code continue working with the slices, since slices would be implementing `FileRegion` properly.
2. Don't have slices extend `FileRegion`, since they don't implement the interface. That'd mean `SliceableFileRegion` would be a completely separate interface from `FileRegion`. Classes could still implement both interfaces though, if they supported both sets of operations.

To me, both of those options seem like fairly small changes to this PR. I'd prefer (1), but I'd be okay with (2).

Which solution do we prefer? (Which may be "3. do nothing; it's great as-is")

Apache9 replied:
The problem of the old `transferTo` method is that, a user can call `thansferTo` multiple times with the same position but the number of `transferred` bytes will be increased every time. Of course this is not the correct way to use `FileRegion` but anyway, it makes it a bit difficult to map the behavior to the new index based solution.

So just follow what I have done to the `DefaultFileRegion` in this PR? Implement two orthogonal transfer APIs? When given a `FileRegion` then you can use the old API, and for `SliceableFileRegion`, you'd better use the new API but it is still OK to use the old API, but it is not allowed to use the old API at first and then switch to the new API, and vice versa. Is it OK? Thanks.

ejona86 replied:
> it makes it a bit difficult to map the behavior to the new index based solution.

I'd generally expect the two APIs to be independent: doing calls on one of the APIs doesn't update the indexes used by the other. I could even accept the two APIs being mutually exclusive: doing calls on one of the APIs breaks the other API (e.g., it might throw an exception).

> When given a FileRegion then you can use the old API, and for SliceableFileRegion, you'd better use the new API but it is still OK to use the old API, but it is not allowed to use the old API at first and then switch to the new API, and vice versa.

Yeah, that sounds fine. I'm not sure I'd go as far as saying it's "not allowed" to mix the APIs, but simply "it makes no sense". The result is the same either way, but the reasoning is a bit more clear.

That's why I think you could use this code, maybe as-is:
```java
private long transferred;

public long transferred() { return transferred; }
￼
public long transferTo(WritableByteChannel target, long position) throws IOException {
  long written = transferBytesTo(target, position, this.count - position);
  if (written > 0) {
    this.transferred += written;
  }
  return written;
}
```

Apache9 replied:
OK, got it. Will provide a new patch soon. Thanks.

Apache9 commented:
@ejona86 I've added the support for `transferTo`. PTAL.

And I didn't remove the related code in `DefaultFileRegion` as we will access the `transferred` field in native transport. I think we can discuss whether we can remove these code in the follow on native transport PR. 

Thanks.

ejona86 commented on transport/src/main/java/io/netty/channel/FileChannelBasedFileRegion.java:
```diff
@@ -0,0 +1,165 @@
+                    + " (expected: 0 - " + (count() - 1) + ')');
+        }
+        if (count == 0) {
+            return 0L;
```
nit: `0`

Apache9 replied:
Done.

Apache9 commented on transport/src/main/java/io/netty/channel/FileChannelBasedFileRegion.java:
```diff
@@ -120,4 +129,37 @@ public SliceableFileRegion transferRetainedSlice(long length) {
+
+    @Override
+    public long transferTo(WritableByteChannel target, long position) throws IOException {
+        // for old style API, the position passed in is not relative, it contains
```
Yeah, misused... We should use `transferred` as the position. Fixed.

ejona86 commented on transport/src/main/java/io/netty/channel/SliceableFileRegion.java:
```diff
@@ -21,8 +21,8 @@
  * <p>
- * We introduce this interface is only because that we can not add new methods to an interface until
- * next minor release. All the methods here will be moved to FileRegion when we release netty 4.2.
+ * We introduce this interface because we we can not add new methods to an interface until next
```
Accidental double "we" :stuck_out_tongue:

Scottmitch commented on transport/src/main/java/io/netty/channel/SliceableFileRegion.java:
```diff
@@ -0,0 +1,131 @@
+     * {@code transferIndex} of this buffer.
+     * <p>
+     * Note that this method returns a {@linkplain #retain() retained} file region unlike
+     * {@link #slice()}.
```
this link is dead -> `{@link #slice(long, long)}`

Apache9 replied:
Done.

Scottmitch commented on transport/src/main/java/io/netty/channel/SliceableFileRegion.java:
```diff
@@ -0,0 +1,131 @@
+     * {@code length}).
+     * <p>
+     * Note that this method returns a {@linkplain #retain() retained} file region unlike
+     * {@link #transferSlice()}.
```
this link is dead -> `{@link #transferSlice(long)}.`

Apache9 replied:
Done.

Scottmitch commented on transport/src/main/java/io/netty/channel/FileChannelBasedFileRegion.java:
```diff
@@ -0,0 +1,161 @@
+
+    @Override
+    public SliceableFileRegion retainedSlice(long index, long length) {
+        SliceableFileRegion sliced = slice(index, length);
```
typically we override the methods which return the same type (e.g. `retain`, `touch`, etc...) so we can do things such as `return slice(index, length).retain();`. Is there any reason why we are not doing that here?

Apache9 replied:
The `retain` and `touch` method in `DefaultFileRegion` return `FileRegion` and this is class is extended by `DefaultFileRegion` so we can not declared a `retain` or `touch` method which returns a `SliceableFileRegion`. And as @normanmaurer said above, users may extend `DefaultFileRegion` so we can not change the return value type of these methods in `DefaultFileRegion`. Maybe we could do this when releasing 4.2?

Scottmitch replied:
thanks for clarifying ... lgtm

Scottmitch commented on transport/src/main/java/io/netty/channel/FileChannelBasedFileRegion.java:
```diff
@@ -0,0 +1,161 @@
+        return sliced;
+    }
+
+    private long transferred;
```
can you move this member variable to the top of the class declaration?

Apache9 replied:
Done.

Scottmitch commented on transport/src/main/java/io/netty/channel/FileChannelBasedSlicedFileRegion.java:
```diff
@@ -0,0 +1,117 @@
+final class FileChannelBasedSlicedFileRegion extends FileChannelBasedFileRegion {
+
+    private final FileChannelBasedFileRegion parent;
+
```
nit: remove extra lines?

Apache9 replied:
Done.

Scottmitch commented on transport/src/main/java/io/netty/channel/FileChannelBasedSlicedFileRegion.java:
```diff
@@ -0,0 +1,117 @@
+    private final long count;
+
+    public FileChannelBasedSlicedFileRegion(FileChannelBasedFileRegion parent, long index,
+            long count) {
```
nit: move onto previous line?

Apache9 replied:
Done.

Scottmitch commented on transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java:
```diff
@@ -206,7 +207,42 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
                     break;
                 }
+            } else if (msg instanceof SliceableFileRegion) {
+                SliceableFileRegion region = (SliceableFileRegion) msg;
```
I wonder if we should break each of these cases into its own private method. Folks may not be using each one of these types and JIT maybe able to optimize better if some unused code is factored out into methods.

@normanmaurer @ejona86 - wdyt?

Apache9 replied:
I'm not sure whether JVM would do better(maybe it will not inline the unused branch and make a smaller doWrite method?) but I think at least it will make the method easier to read and understand. I can do the refactoring if no objections. Thanks.

Scottmitch replied:
refactoring out not-commonly called code paths can reduce the byte code size of a method and make it easier to inline ... we don't have todo that as part of this PR but in theory seems like a good thing todo.

Scottmitch commented on transport/src/test/java/io/netty/channel/FileChannelBasedSlicedFileRegionTest.java:
```diff
@@ -0,0 +1,174 @@
+
+    private byte[] transferBytesTo(SliceableFileRegion region, int length) throws IOException {
+        ByteArrayOutputStream bos = new ByteArrayOutputStream();
+        WritableByteChannel ch = Channels.newChannel(bos);
```
does `ch` need to be closed?

Apache9 replied:
Not really, it just wraps a `ByteArrayOutputStream`. But anyway, no harm to close it. Fixed with a try finally.

Scottmitch commented on transport/src/test/java/io/netty/channel/FileChannelBasedSlicedFileRegionTest.java:
```diff
@@ -0,0 +1,174 @@
+
+    private byte[] transferTo(FileRegion region) throws IOException {
+        ByteArrayOutputStream bos = new ByteArrayOutputStream();
+        WritableByteChannel ch = Channels.newChannel(bos);
```
does `ch` need to be closed?

Apache9 replied:
Ditto.

Apache9 commented:
@Scottmitch Filed #7155 for the follow on work. Do not know why I can not reply to the comment directly...

fenik17 commented on transport/src/main/java/io/netty/channel/FileChannelBasedFileRegion.java:
```diff
@@ -0,0 +1,160 @@
+        long count = count() - position;
+        if (count < 0 || position < 0) {
+            throw new IllegalArgumentException("position out of range: " + position
+                    + " (expected: 0 - " + (count() - 1) + ')');
```
@carl-mastrangelo one of the IntelliJ IDEA's inspections: "Reports String literals of length one being used in concatenation. These literals may be replaced by equivalent character literals, gaining some performance enhancement."

See https://stackoverflow.com/questions/24859500/concatenate-char-literal-x-vs-single-char-string-literal-x

Apache9 replied:
So let's keep it as a character? The code is just the same with `DefaultFileRegion.transferTo`.

carl-mastrangelo replied:
@fenik17 if the reason for the char is to silence IDE warnings, I get that.  If it's a perf thing, this code path is unlikely to be executed a lot.  That would not be a good reason.

Personally I prefer format strings as I find them easier to read.

fenik17 replied:
Yes, at this place it is just to suppress IDE warnings and to the be consistence with Netty's code style. 
+1 for use `String.format()` for the exceptions messages.

Apache9 replied:
Done.

netkins commented on transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java:
```diff
@@ -208,36 +268,19 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
-                } else {
-                    // Break the loop and so incompleteWrite(...) is called.
-                    break;
+                switch (doWriteFileRegion(in, region, writeSpinCount)) {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a default case to this switch. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)

netkins commented on transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java:
```diff
@@ -157,6 +158,65 @@ public final void read() {
+            in.progress(flushedAmount);
+            return status;
+        } else {
+            // TODO: consider remove this branch once we move the methods in SliceableFileRegion
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

Apache9 commented on transport/src/main/java/io/netty/channel/FileChannelBasedFileRegion.java:
```diff
@@ -0,0 +1,160 @@
+            throws IOException {
+        if (position < 0 || position > count()) {
+            throw new IllegalArgumentException("position out of range: " + position
+                    + " (expected: 0 - " + (this.count() - 1) + ')');
```
Done.

Apache9 commented on transport/src/main/java/io/netty/channel/FileChannelBasedFileRegion.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    /**
+     * Explicitly open the underlying file-descriptor if not done yet.
```
Done.

Apache9 commented:
Any other concerns? @Scottmitch @carl-mastrangelo . Thanks.

carl-mastrangelo commented on transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java:
```diff
@@ -157,12 +158,71 @@ public final void read() {
+            WriteStatus status = WriteStatus.INCOMPLETE;
+            long flushedAmount = 0;
+
+            for (int i = writeSpinCount - 1; i >= 0; i--) {
```
Why not use a plain incrementing for loop?

Scottmitch replied:
this is preexisting IIRC ... the index isn't used and we don't have to declare an additional constant for `end` this way.

carl-mastrangelo commented on transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java:
```diff
@@ -157,12 +158,71 @@ public final void read() {
 
         boolean setOpWrite = false;
-        for (;;) {
+        LOOP: for (;;) {
```
Upper case should be reserved for constants.  is a label really a constant?

Apache9 replied:
This is only my style...

What do we suggest in netty? 'Loop'? 'loop'?

normanmaurer replied:
`loop:`

Apache9 replied:
Done.

Apache9 commented on transport/src/main/java/io/netty/channel/FileChannelBasedFileRegion.java:
```diff
@@ -0,0 +1,162 @@
+        // Call open to make sure fc is initialized. This is a no-op if we called it before.
+        open();
+
+        long written = channel().transferTo(position() + position, count, target);
```
The reference count check is in the `channel()` method... The method used to be public so we need to do reference count check. Let me do some refactoring here.

Apache9 replied:
Make `transferTo` call `transferBytesTo`, add ref count check.

Apache9 commented:
Any more concerns?  Thanks.

Apache9 commented:
Ping @normanmaurer  @Scottmitch

Scottmitch commented on transport/src/main/java/io/netty/channel/FileChannelBasedFileRegion.java:
```diff
@@ -0,0 +1,153 @@
+
+    @Override
+    public long transferBytesTo(WritableByteChannel target, long position, long length)
+            throws IOException {
```
nit: put this on the previous line

Apache9 replied:
Done. So what's is expected line length? 120?

normanmaurer replied:
yes... 120

Apache9 commented:
@Scottmitch  Rebase and squash done.

Scottmitch commented:
@normanmaurer - Please pull this in after the next release if you are satisfied.

Apache9 commented:
4.1.16 is out, is it OK to pull in this now? There are several follow on issues, such as adding native transport support, and also need to start trying to add FileRegion support for HTTP/2.

Thanks.

normanmaurer commented:
@Apache9 I will try to have a final look this week

Apache9 commented:
Any updates here? @normanmaurer 
Thanks.

Scottmitch commented:
@Apache9 - sorry for delay ... thanks for your patience ... hang in there we will pull this in soon.

Apache9 commented:
Thanks @Scottmitch  . Wait for your good new.

Apache9 commented:
Ping @Scottmitch @normanmaurer .

Thanks.

normanmaurer commented:
@Apache9 sorry I know you wait forever :( Its on my to do list for this week.

Apache9 commented:
Wait for your good news.

normanmaurer commented:
@Apache9 sorry if I asked this before but can you please tell me again what exactly you need this for ? Is it "just" so you can write a File in slices without the need to open multiple FileChannels ?

Apache9 commented:
@normanmaurer We need this to add the FileRegion support for HTTP/2. We can not embed a large FileRegion in a single data frame when we actually write it out, we need to split it to several small ones.

Thanks.

normanmaurer commented:
@Apache9 ok so its related to what I thought... Now a maybe stupid question but why could you not just open a FileChannel and create multiple new DefaultFileRegion using this FileChannel with different position and count arguments ? The only "problem" I can see here is that you need to ensure you not close the FileChannel under the hood when not all are written yet.

Apache9 commented:
So you mean we should expose an API in HTTP/2 with something like a FileRegion but not a FileRegion? Maybe Http2FileRegion? Which will have the slice support and returns a FileRegion for slice? I'm OK with it. Just want to know your decision.

Thanks.

normanmaurer commented:
@Apache9 why you can't just write a FileRegion with the correct size from the beginning ? Or you want to "enqueue" something into the HTTP2 layer and dynamically slice out the correct stuff ?

Apache9 commented:
Oh, maybe I need to say more. For the end users of the HTTP/2 API, it is impossible for them to write a `FileRegion` with exactly the correct size for a data frame, there are flow controls, and there maybe others writing on the same connection but different streams, so the next data frame available to them will be changed dynamically. In the framework we will do splitting and merging of data frames.

So I think the correct way to do this is to give end users the ability to write a very large `FileRegion` or something like a `FileRegion`, and we will split it to several small data frames in the framework.

Thanks.

normanmaurer commented:
@Apache9 ok got it...

So I think what I would do is to have another abstraction which basically provides a "Stream" of FileRegion on demand. I think this would minimise the changeset a lot. I am not convinced that we really need the slice stuff on the FileRegion itself.

Let me open a PR that shows the idea... Gimme a few minutes ;)

normanmaurer commented:
@Apache9 check https://github.com/netty/netty/pull/7435 ... No unit tests included yet ;)

normanmaurer commented:
@Apache9 sorry for all the delay and back and forth by @Scottmitch and myself were discussion some alternatives approves and at the end we think we may want to mark `FileRegion` as `@deprecated` and introduce some new API. Hopefully I can write up some proposal soon.

Apache9 commented:
@normanmaurer Thanks. Wait for your good news.

normanmaurer commented:
Let me close this as this is not really something we want to do for 4.1.
